modules/extensions/src/terrain/terrain-cover.ts

```diff
@@ -0,0 +1,251 @@
+      const leftBottomCommon = owner.projectPosition(newBounds[0], {viewport});
+      const topRightCommon = owner.projectPosition(newBounds[1], {viewport});
+      this.commonBounds = [leftBottomCommon, topRightCommon];
+    } else {
+      this.commonBounds = null;
+    }
+
+    if (needsRedraw) {
+      this.renderViewport = getRenderViewport(owner, this.zoom, viewport.isGeospatial);
+    }
+    return needsRedraw;
+  }
+
+  get renderTexture(): Framebuffer | null {
+    if (!this.renderViewport || this.layers.length === 0) {
+      return null;
+    }
+    if (!this.fbo) {
+      this.fbo = createRenderTarget(this.owner.context.gl, {id: this.id});
+    }
+    return this.fbo;
+  }
+
+  get pickingTexture(): Framebuffer | null {
+    if (!this.renderViewport || this.layers.length === 0) {
+      return null;
+    }
+    if (!this.pickingFbo) {
+      this.pickingFbo = createRenderTarget(this.owner.context.gl, {id: `${this.id}-picking`});
+    }
+    return this.pickingFbo;
+  }
+
+  filterLayers(layers: Layer[]) {
+    return layers.filter(({id}) => this.layers.includes(id));
+  }
+
+  delete() {
+    const {fbo, pickingFbo} = this;
+    if (fbo) {
+      fbo.texture.delete();
+      fbo.delete();
+    }
+    if (pickingFbo) {
+      pickingFbo.texture.delete();
+      pickingFbo.delete();
+    }
+  }
+}
+
+function getRenderViewport(layer: Layer, zoom: number, isGeospatial: boolean): Viewport | null {
+  // Construct a viewport that just covers the owner layer's bounds
+  const bounds = layer.getBounds();
+  if (!bounds) {
+    return null;
+  }
+
+  const refViewport: Viewport = isGeospatial ? unitGeospatialViewport : unitNonGeospatialViewport;
+  const leftBottomCommon = layer.projectPosition(bounds[0], {viewport: refViewport});
+  const topRightCommon = layer.projectPosition(bounds[1], {viewport: refViewport});
+
+  const center = refViewport.unprojectPosition([
+    (leftBottomCommon[0] + topRightCommon[0]) / 2,
+    (leftBottomCommon[1] + topRightCommon[1]) / 2,
+    0
+  ]);
+
+  const scale = 2 ** zoom;
+  let width = Math.round(Math.abs(topRightCommon[0] - leftBottomCommon[0]) * scale);
+  let height = Math.round(Math.abs(topRightCommon[1] - leftBottomCommon[1]) * scale);
+  if (width > MAX_SIZE || height > MAX_SIZE) {
+    const r = MAX_SIZE / Math.max(width, height);
+    width = Math.round(width * r);
+    height = Math.round(height * r);
+  }
+
+  return isGeospatial
+    ? new WebMercatorViewport({
+        width,
+        height,
+        longitude: center[0],
+        latitude: center[1],
+        zoom,
+        orthographic: true
+      })
+    : new OrthographicViewport({width, height, target: center, zoom, flipY: false});
+}
+
+/**
+ * Remove layers that do not overlap with the current terrain cover.
+ * This implementation only has effect when a TileLayer is overlaid on top of a TileLayer
+ */
+function getIntersectingLayers(sourceTile: Tile2DHeader | null, layers: Layer[]): Layer[] {
+  if (sourceTile) {
```

Comment:
Nit: Why support a null tile here? It kind of dilutes the typing and clear purpose.